### ISSUE TITLE
Rule proposal: no-unsafe-callback 

### ISSUE THREAD

**Issue by lgenzelis** [typescript-eslint/typescript-eslint#4300]
### Description

**Disclaimer:** I posted this as [a question in SO](https://stackoverflow.com/questions/70335232/typescript-eslint-rule-that-prevents-passing-a-callback-with-a-typed-argument-to), but it seems that the implicit answer is "we need a new rule".

Let's suppose an external library has this code:

```typescript
type Callback = (x: any) => void;

function invokesCallback(callback: Callback): void {
  callback({ thisWill: 'break everything' });
}
```
and then, in our codebase, we have this:

```typescript
const someNumbers: number[] = [];
function appendNumber(value: number) {
  someNumbers.push(value);
}

invokesCallback(appendNumber);
```

Even with all the `@typescript-eslint/no-unsafe-*` rules in my `.eslintrc.js` file, eslint doesn't even warn me about the issue.

### Fail

Let's asume an external module with these declarations:
```ts
type Callback1 = (x: any) => void;
type Callback2 = (x: any, y: any) => void;
type Callback3 = ({ x, y }: { x: any, y: string}) => void;

declare function invokesCallback1(callback: Callback1): void;
declare function invokesCallback2(callback: Callback2): void;
declare function invokesCallback3(callback: Callback3): void;
```

Then, these should fail:
```ts
type MyCallback1 = (x: string) => void;
type MyCallback2 = (x: number, y: string) => void;
type MyCallback3 = ({ x, y }: { x: number, y: string}) => void;

declare const myCallback1: MyCallback1;
declare const myCallback2: MyCallback2;
declare const myCallback3: MyCallback3;

invokesCallback1(myCallback1);
invokesCallback2(myCallback2);
invokesCallback3(myCallback3);
```

### Pass

Assuming the same external declarations as before, these should pass:
```ts
type MyCallback1 = (x: unknown) => void;
type MyCallback2 = (x: unknown) => void;
type MyCallback3 = ({ x, y }: { x: unknown, y: string}) => void;

declare const myCallback1: MyCallback1;
declare const myCallback2: MyCallback2;
declare const myCallback3: MyCallback3;

invokesCallback1(myCallback1);
invokesCallback2(myCallback2);
invokesCallback3(myCallback3);
```

**Comment by bradzacher** [comment]
The difference with this request in comparison to the `no-unsafe-*` rules is that those rules [consider an assignment from a type ***TO*** `any` as safe](https://typescript-eslint.io/play/#ts=4.5.2&sourceType=module&code=PQKgBApgzgNglgOwC5gAJIJ4AdoGMBOcWSAtNPMsAgPYkCuCUAhgGYQlNRRwDmCAthGQAuMAHII+fNXxiwIYAFgAUABMIuGE3wQwuaoxQAPUUwQYA3CpX7DYDKIR1+AI0lgAvGCMWwwYJBSMgA0YJxhXLwIiDxh5mBI1GBOru5wUGAMzGw2BlAoAF6m8V6WfgE0gdL4oeGc3Hwxyc5u+AlJZhhg6WAARNkQvUA&rules=N4XyA&tsConfig=N4XyA).

This request you want to treat an assignment from a type to `any` to be unsafe, which is the inverse check.
Funnily though, this is actually correctly inverted for functions: it is "safe" to assign a function type which has an `any`-typed parameter to a location accepting a non-`any`-typed parameter, but it is unsafe to do the inverse.

This would ultimately be very complex to write - inspecting function arguments isn't the easiest thing in TS due to how the type system works.
Additionally there are valid cases we need to take care of.

For example: `function foo<T extends (...any[]) => void>(fn: T): T` is the only valid way of declaring a function which accepts any function and returns a function of the same type.
For example - in react the `useCallback` hook (which memoizes the callback function) uses exactly this signature.
So the logic would need to handle these sorts of valid cases, or else codebases that use these patterns would be littered with false-positives.

----

As an aside - the best course of action is to use a tool like [`patch-package`](https://www.npmjs.com/package/patch-package) to locally fix the types to use `unknown` (or just the correct type), and then PR the package to fix the types to be sound.

**Comment by lgenzelis** [comment]
Thank you for your suggestion @bradzacher! Honestly, I've got no idea how rules are written, so I can't say anything related to how hard it'd be to create this one 😕 

I'm curious about the valid use cases you mention

> `function foo<T extends (...any[]) => void>(fn: T): T` is the only valid way of declaring a function which accepts any function and returns a function of the same type.

Wouldn't `function foo<T extends (...unknown[]) => void>(fn: T): T` achieve the exact same thing?

**Comment by bradzacher** [comment]
It would, now. But many, many libraries still use the `any[]` form as it was the only way to do it for a good long while in TS.

As mentioned - react's `useCallback` is one of the widest-used examples of this.

**Comment by lgenzelis** [comment]
> many libraries still use the any[] form as it was the only way to do it for a good long while in TS

Then it would be up to the users to decide whether to enable this rule or not, and if they do, it'd be up to them to disable it for things they believe are actually safe (like `useCallback `). That's my opinion, though.

**Comment by raveclassic** [comment]
Another pretty common use case for such rule would be callbacks passed to `Promise.catch`:
```typescript
Promise.reject(123).catch((value: string) => {
  console.log(value.toUpperCase()) // WUT???
})
```
